coap/api/transport.go

```diff
@@ -66,75 +67,78 @@ func MakeCoAPHandler(svc coap.Service, l *slog.Logger) mux.HandlerFunc {
 	return handler
 }
 
-func sendResp(w mux.ResponseWriter, resp *message.Message) {
-	if err := w.Client().WriteMessage(resp); err != nil {
+func sendResp(w mux.ResponseWriter, resp *pool.Message) {
+	if err := w.Conn().WriteMessage(resp); err != nil {
 		logger.Warn(fmt.Sprintf("Can't set response: %s", err))
 	}
 }
 
 func handler(w mux.ResponseWriter, m *mux.Message) {
-	resp := message.Message{
-		Code:    codes.Content,
-		Token:   m.Token,
-		Context: m.Context,
-		Options: make(message.Options, 0, 16),
+	resp := pool.NewMessage(w.Conn().Context())
+	resp.SetToken(m.Token())
+	for _, opt := range m.Options() {
+		resp.AddOptionBytes(opt.ID, opt.Value)
 	}
-	defer sendResp(w, &resp)
+	defer sendResp(w, resp)
+
 	msg, err := decodeMessage(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error decoding message: %s", err))
-		resp.Code = codes.BadRequest
+		resp.SetCode(codes.BadRequest)
 		return
 	}
 	key, err := parseKey(m)
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error parsing auth: %s", err))
-		resp.Code = codes.Unauthorized
+		resp.SetCode(codes.Unauthorized)
 		return
 	}
-	switch m.Code {
+
+	switch m.Code() {
 	case codes.GET:
-		err = handleGet(m.Context, m, w.Client(), msg, key)
+		err = handleGet(m.Context(), m, w, msg, key)
+		resp.SetCode(codes.Content)
 	case codes.POST:
-		resp.Code = codes.Created
-		err = service.Publish(m.Context, key, msg)
+		resp.SetCode(codes.Created)
+		err = service.Publish(m.Context(), key, msg)
 	default:
 		err = svcerr.ErrNotFound
 	}
+
 	if err != nil {
 		switch {
 		case err == errBadOptions:
-			resp.Code = codes.BadOption
+			resp.SetCode(codes.BadOption)
 		case err == svcerr.ErrNotFound:
-			resp.Code = codes.NotFound
+			resp.SetCode(codes.NotFound)
 		case errors.Contains(err, svcerr.ErrAuthorization),
 			errors.Contains(err, svcerr.ErrAuthentication):
-			resp.Code = codes.Unauthorized
+			resp.SetCode(codes.Unauthorized)
 		default:
-			resp.Code = codes.InternalServerError
+			resp.SetCode(codes.InternalServerError)
 		}
 	}
 }
 
-func handleGet(ctx context.Context, m *mux.Message, c mux.Client, msg *messaging.Message, key string) error {
+func handleGet(ctx context.Context, m *mux.Message, w mux.ResponseWriter, msg *messaging.Message, key string) error {
 	var obs uint32
-	obs, err := m.Options.Observe()
+	obs, err := m.Options().Observe()
 	if err != nil {
 		logger.Warn(fmt.Sprintf("Error reading observe option: %s", err))
 		return errBadOptions
 	}
 	if obs == startObserve {
-		c := coap.NewClient(c, m.Token, logger)
+		c := coap.NewClient(w.Conn(), m.Token(), logger)
 		return service.Subscribe(ctx, key, msg.GetChannel(), msg.GetSubtopic(), c)
 	}
-	return service.Unsubscribe(ctx, key, msg.GetChannel(), msg.GetSubtopic(), m.Token.String())
+	return service.Unsubscribe(ctx, key, msg.GetChannel(), msg.GetSubtopic(), m.Token().String())
 }
```

Comment:
suggestion to fix error mentioned at here https://github.com/absmach/magistrala/pull/2183#pullrequestreview-2020116831

```suggestion
func handleGet(m *mux.Message, w mux.ResponseWriter, msg *messaging.Message, key string) error {
	var obs uint32
	obs, err := m.Options().Observe()
	if err != nil {
		logger.Warn(fmt.Sprintf("Error reading observe option: %s", err))
		return errBadOptions
	}
	if obs == startObserve {
		c := coap.NewClient(w.Conn(), m.Token(), logger)
		return service.Subscribe(w.Conn().Context(), key, msg.GetChannel(), msg.GetSubtopic(), c)
	}
	return service.Unsubscribe(w.Conn().Context(), key, msg.GetChannel(), msg.GetSubtopic(), m.Token().String())
}
```